src/almanack/processing/compute_data.py

```diff
@@ -81,6 +81,86 @@ def compute_repo_data(repo_path: str) -> None:
         return {"repo_path": str(repo_path), "error": str(e)}
 
 
+from typing import Any, Dict
+
+
+def compute_pr_data(repo_path: str, pr_branch: str, main_branch: str) -> Dict[str, Any]:
```

Comment:
This function's body seems awfully similar to `compute_repo_data()` above; perhaps you could add `most_recent_commit` and `oldest_commit` parameters to `compute_repo_data()`, with it defaulting to the first and last commit if unspecified, and then use that in `compute_pr_data()`?

I'm thinking you'd modify `compute_repo_data()` like so:

```python
def compute_repo_data(repo_path: str, most_recent_commit: pygit2.Commit=None, oldest_commit: pygit2.Commit=None, ) -> None:
    # ...
    
    # Retrieve the list of commits from the repository
        commits = get_commits(repo)
        most_recent_commit = commits[0] if most_recent_commit is None else most_recent_commit
        first_commit = commits[-1] if most_recent_commit is None else most_recent_commit
```

Then, you can invoke it from `compute_pr_data()` like so:

```python
def compute_pr_data(repo_path: str, pr_branch: str, main_branch: str) -> Dict[str, Any]:
    try:
        # ...
    
        # Get the most recent commits on each branch
        pr_commit = repo.get(pr_ref.target)
        main_commit = repo.get(main_ref.target)
    
        result = compute_repo_data(repo_path, pr_commit, main_commit)
        
        return {
            "pr_branch": pr_branch,
            "main_branch": main_branch,
            "total_entropy_introduced": result["total_normalized_entropy"],
            "number_of_files_changed": result["number_of_files"],
            "entropy_per_file": result["file_level_entropy"],
            "commits": result["time_range_of_commits"]
        }
        
      except Exception as e:
          # If processing fails, return an informative error
          return {"pr_branch": pr_branch, "main_branch": main_branch, "error": str(e)}
```

There is a small downside in that you have to parse the repo twice; IMHO that isn't a big deal, but if you want to avoid that, there are a number of ways you could do it. You could, for example, have `compute_repo_data()` take a `repo` object, then write another small method to deal with taking in a path and constructing `repo` object, which would then call `compute_repo_data(repo)`.  Alternatively, you could pull the entropy calculations out of `compute_repo_data()` into another function and call it from both `compute_repo_data()` and `compute_pr_data()`.

I'd also suggest using consistent names with your variables and dictionary keys. For example, in `compute_repo_data()`, you use `normalized_total_entropy` as a variable name, then assign it to the dict key `total_normalized_entropy`. Another example: in `compute_pr_data()` you call the file-level entropy in the results dict `entropy_per_file`, but in `compute_repo_data` it's `file_level_entropy`.

Reply:
Thanks for this comment @falquaddoomi! I definitely agree that there is some serious overlap between those two functions, and this refactoring is needed. Going to add this to a new issue. I got things running but ran into some small troubles with test cases and other small errors. With such a small deadline before my presentation, I figure it makes more sense to save this for later on.